chartlets.js/src/lib/components/Plot.tsx

```diff
@@ -2,32 +2,109 @@ import { VegaLite } from "react-vega";
 
 import { type PlotState } from "@/lib/types/state/component";
 import { type ComponentChangeHandler } from "@/lib/types/state/event";
+import {
+  isString,
+  isTopLevelSelectionParameter,
+  type SignalHandler,
+} from "@/lib/types/state/vega";
+import { useCallback, useMemo } from "react";
 
 export interface PlotProps extends Omit<PlotState, "type"> {
   onChange: ComponentChangeHandler;
 }
 
 export function Plot({ id, style, chart, onChange }: PlotProps) {
```

Comment:
There is a lot of logic now in an UI component. Please extract logic into a dedicated hook, the use it like so:

```typescript
  const signalListeners = useSignalListeners(id, onChange);
```